packages/fast-usdc/src/exos/status-manager.js

```diff
@@ -1,12 +1,152 @@
+import { M } from '@endo/patterns';
+import { makeError, q } from '@endo/errors';
+
+import { CctpTxEvidenceShape, StatusManagerEntryShape } from '../typeGuards.js';
+import { TxStatus } from '../constants.js';
+
 /**
  * @import {Zone} from '@agoric/zone';
+ * @import {CctpTxEvidence, NobleAddress, StatusManagerEntry} from '../types.js';
  */
 
+/** @param {CctpTxEvidence} evidence */
+export const getStatusKey = evidence => {
+  const {
+    tx: { amount, forwardingAddress },
+  } = evidence;
+  return `${forwardingAddress}-${amount}`;
+};
+
 /**
  * @param {Zone} zone
  */
 export const prepareStatusManager = zone => {
-  return zone.exo('Fast USDC Status Manager', undefined, {});
+  const txs = zone.mapStore('txs', {
+    keyShape: M.string(),
+    valueShape: M.arrayOf(StatusManagerEntryShape),
+  });
+
+  return zone.exo(
+    'Fast USDC Status Manager',
+    M.interface('StatusManagerI', {
+      observe: M.call(CctpTxEvidenceShape).returns(M.number()),
+      advance: M.call(M.string(), M.bigint(), M.number()).returns(
+        M.undefined(),
+      ),
+      hasPendingSettlement: M.call(M.string(), M.bigint()).returns(M.boolean()),
+      settle: M.call(M.string(), M.bigint()).returns(M.undefined()),
+      view: M.call(M.string(), M.bigint())
+        .optional(M.number())
+        .returns(
+          M.or(
+            StatusManagerEntryShape,
+            M.arrayOf(StatusManagerEntryShape),
+            M.undefined(),
+          ),
+        ),
+    }),
+    {
+      /**
+       * Queue a transaction before initiating a transfer. Receive
+       * the entry index in return.
+       *
+       * @param {CctpTxEvidence} evidence
+       * @returns {number}
+       */
+      observe(evidence) {
+        // XXX like `appendToStoredArray`, but returns index. consider
+        // upstreaming the change if it doesn't affect existing consumers
+        const key = getStatusKey(evidence);
+        const entry = { ...evidence, status: TxStatus.OBSERVED };
+        if (!txs.has(key)) {
+          // XXX append to store helper
+          txs.init(key, harden([entry]));
+          return 0;
+        }
+        const current = txs.get(key);
+        // XXX do we need defensive logic to ensure multiple entries don't have
+        // the same txHash+chainId?
+        txs.set(key, harden([...current, entry]));
+        return current.length;
+      },
+      /**
+       * Mark an `OBSERVED` transaction as `ADVANCED`.
+       *
+       * @param {string} address
+       * @param {bigint} amount
+       * @param {number} index
+       */
+      advance(address, amount, index) {
+        const key = `${address}-${amount}`;
+        const mutableEntries = [...txs.get(key)];
+        if (mutableEntries[index].status !== TxStatus.OBSERVED) {
+          throw makeError(
+            `Invalid status ${q(mutableEntries[index].status)} for ${q(key)}`,
+          );
+        }
+        mutableEntries[index] = {
+          ...mutableEntries[index],
```

Comment:
Iterating over `mutableEntries` is doing O(n) work for an O(1) size message, which is against [the rules](https://github.com/Agoric/agoric-sdk/wiki/Limit-work-by-message-size).

Please add at least a comment explaining why it's OK.